File: 1023-time-based-key-value-store/time-based-key-value-store.java
import java.util.Vector;
import java.util.HashMap;

class TimeMap {
    HashMap<String, Vector<Pair<String, Integer>>> time = new HashMap<String, Vector<Pair<String, Integer>>>();

    public TimeMap() {
        
    }
    
    public void set(String key, String value, int timestamp) {
        if(!time.containsKey(key)){
            Vector<Pair<String, Integer>> temp = new Vector<Pair<String, Integer>>();
            time.put(key, temp);
        }
        Pair<String, Integer> temp = new Pair<String, Integer>(value, timestamp);
        Vector<Pair<String, Integer>> temp2 = time.get(key);
        temp2.add(temp);
        time.put(key, temp2);
    }
    
    public String get(String key, int timestamp) {
        String result = "";
        if(!time.containsKey(key)) return result;
        Vector<Pair<String, Integer>> times = time.get(key);
        int l = 0;
        int r = times.size() - 1;
        while(l<=r){
            int m = l + ((r - l) / 2);

            if(times.get(m).getValue() == timestamp){
                result = times.get(m).getKey();
                break;
            }
            else if(times.get(m).getValue() < timestamp){
                result = times.get(m).getKey();
                l = m + 1;
            }
            else{
                r= m - 1;
            }
        }
        return result;
    }
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap obj = new TimeMap();
 * obj.set(key,value,timestamp);
 * String param_2 = obj.get(key,timestamp);
 */

File: 1023-time-based-key-value-store/time-based-key-value-store.cpp
#include<unordered_map>
class TimeMap {
private:
    unordered_map<string, vector<pair<int, string>>> timemap;
public:
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        timemap[key].emplace_back(timestamp, value);
    }
    
    string get(string key, int timestamp) {
        string result = "";

        auto& timestamps = timemap[key];

        int l = 0, r = timestamps.size() - 1;

        while(l <= r){
            int m = l + ((r - l) / 2);
            if(timestamps[m].first <= timestamp){
                result = timestamps[m].second;
                l = m + 1; 
            }
            else r = m - 1;
            
        }
        return result;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */